shopfloor/models/stock_move_line.py

```diff
@@ -21,3 +21,86 @@ class StockMoveLine(models.Model):
 
     # allow domain on picking_id.xxx without too much perf penalty
     picking_id = fields.Many2one(auto_join=True)
+
+    def _split_pickings_from_source_location(self):
+        """Ensure that the related pickings will have the same source location.
+
+        Some pickings related could have other unrelated move lines, as such we
+        have to split them to contain only the move lines related to the expected
+        source location.
+
+        Example:
+
+            Initial data:
+
+                PICK1:
+                    - move line with source location LOC1
+                    - move line with source location LOC2
+                PICK2:
+                    - move line with source location LOC2
+                    - move line with source location LOC3
+
+            Then we process move lines related to LOC2 with this method, we get:
+
+                PICK1:
+                    - move line with source location LOC1
+                PICK2:
+                    - move line with source location LOC3
+                PICK3:
+                    - move line with source location LOC2
+                    - move line with source location LOC2
+
+        Return the new picking (in case a split has been made), or the current
+        related pickings.
+        """
+        location_src_to_process = self.location_id
+        if location_src_to_process:
+            assert (
+                len(location_src_to_process) == 1
+            ), "Move lines processed have to share the same source location."
+        pickings = self.picking_id
+        move_lines_to_process_ids = []
+        for picking in pickings:
+            location_src = picking.mapped("move_line_ids.location_id")
+            if len(location_src) == 1:
+                continue
+            # Get the related move lines among the picking and split them
+            move_lines_to_process_ids.extend((picking.move_line_ids & self).ids)
```

Comment:
Updated!
Often I prefer to preserve the benefit of the object types to avoid the 'mixing apples and oranges' issues, but I understand it could add some overhead...